🐛 clusterctl: fix goproxy to also return versions for major > 1

chrischdi opened the pull request:
<!-- please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:

This PR fixes the goproxy client to also list versions of go modules > v1.
The previously used goproxy endpoint only lists versions of major v1.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7700

**Additional info**:

The goproxy client will not provide the correct results for go based provider which do not follow go module semantical versioning (e.g. tagging v2.0.0 but not setting go.mod to `/v2`. [xref](https://github.com/kubernetes-sigs/cluster-api/issues/7700#issuecomment-1340759006). If this is considered as a no-go, I would propose to remove the goproxy implementation instead of fixing it. Such providers are always able to set `GOPROXY=off` to allow clusterctl to directly use the github api.

vincepri reviewed (approved):
/lgtm

Should we also fix the code in https://github.com/kubernetes-sigs/cluster-api/blob/main/hack/tools/mdbook/releaselink/releaselink.go ?

We should probably dedupe these two and favor less copying around, wdyt?

chrischdi commented:
> /lgtm
> 
> Should we also fix the code in https://github.com/kubernetes-sigs/cluster-api/blob/main/hack/tools/mdbook/releaselink/releaselink.go ?
> 
> We should probably dedupe these two and favor less copying around, wdyt?

Great point. Did already forget that we've "stolen" that from there :-)
Also adjusted the parameter for aws to show the latest version in the doc.

I moved it to `internal/goproxy` for now. Happy to get better recommendations 👍 
I also added it as separate commit for now to better show the diff to the first commit. Will squash later on.

ykakarap commented on internal/goproxy/goproxy_test.go:
```diff
@@ -0,0 +1,210 @@
+
+			got, err := clientGoproxy.GetVersions(ctx, tt.gomodulePath)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("Client.GetVersions() error = %v, wantErr %v", err, tt.wantErr)
```
How about using gomega assertions? 
We use gomega everywhere else in the project. It would be nice to have consistency.

(Same comment at other places in the file).

ykakarap commented on internal/goproxy/goproxy.go:
```diff
@@ -0,0 +1,192 @@
+}
+
+// NewGoproxyClient returns a new goproxyClient instance.
+func NewGoproxyClient(scheme, host string) *Client {
```
```suggestion
func NewClient(scheme, host string) *Client {
```
nit: Since the package is already called `goproxy` it new function will read `goproxy.NewClient()`.

ykakarap commented on hack/tools/mdbook/releaselink/releaselink.go:
```diff
@@ -58,36 +56,22 @@ func (l ReleaseLink) Process(input *plugin.Input) error {
 		includePrereleases := tags.Get("prereleases") == "true"
 
-		repo, err := vcs.RepoRootForImportPath(gomodule, false)
+		scheme, host, err := goproxy.GetSchemeAndHost(os.Getenv("GOPROXY"))
```
Since now we are reading from `GOPROXY` env variable we should account for `scheme` and `host` being empty strings. 

If `GOPROXY` envvar is set to `direct` or `off` we will get an empty `scheme` and `host` and the code below will fail.

chrischdi replied:
Added returning an error when one of scheme or host is empty to fail.

ykakarap reviewed (changes requested):
The changes to support >v1 versions looks good.

There is an edge case where this might now work. Example when a go modules jumps from v1 to v3 then the current code will not list the v3 versions as it will stop immediately after v2 is not found. This is a very rare case and I dont know think why someone might want to jump a major version. But, I thought I would point this out anyway. 

FYI: This is a breaking change in clusterctl's library but given our [compatibility notice](https://github.com/kubernetes-sigs/cluster-api/tree/main/cmd/clusterctl#compatibility-notice) for clusterctl we should be fine and no further action is needed.

chrischdi commented:
Thanks for reviewiewing.

> The changes to support >v1 versions looks good.
> 
> There is an edge case where this might now work. Example when a go modules jumps from v1 to v3 then the current code will not list the v3 versions as it will stop immediately after v2 is not found. This is a very rare case and I dont know think why someone might want to jump a major version. But, I thought I would point this out anyway.

Yes great point! That's why I mentioned it in the office hours and also added the note at the PR description 👍

chrischdi commented:
/test pull-cluster-api-e2e-informing-main

chrischdi commented:
Squashed both commits to a single one.

killianmuldoon reviewed (commented):
Looks good - just a couple nits

fabriziopandini commented:
Thanks @chrischdi for quicky jumping on this issue!
/lgtm
/approve

chrischdi commented:
/cherry-pick release-1.3
